### PR TITLE
Generate BuildConfig headers in GN

### PR DESCRIPTION
**Pull request by mspang**
Generate config headers to save the configuration options to be installed
or used by another build system.

Unlike autoconf but like the main config headers (e.g. CHIPConfig.h)
this generates separate configs per component (e.g. CHIPBuildConfig.h).
This will allow scoping of options to particular config headers, which
didn't quite work properly with automake (options set by autoconf became
available everywhere).

The header generator template is derived from Chrome's, but with some
extra wrapping of define usage with BUILDFLAG() removed. That's a safety
feature that we may want to bring back - it triggers an error when macros
resolve to zero after forgetting an #include. Alternatively, we can enable
-Wundef, although all integrations may not be able to do that.

This is a prerequisite for closing #1513.

### CODE REVIEW

File: examples/lighting-app/nrf5/BUILD.gn [hunk 1]
```diff
@@ -14,6 +14,7 @@
 
 import("//build_overrides/chip.gni")
 import("//build_overrides/nrf5_sdk.gni")
+import("//build_overrides/openthread.gni")
```
Comment (rwalker-apple) [inline examples/lighting-app/nrf5/BUILD.gn:17]: is this part relevant to BuildConfig.h?

**Review by rwalker-apple — APPROVED** [review]
lgtm.

This would have saved me a lot of work in my recent "gn in chip_xcode_connector.sh" work, but I'm a little concerned longer term CHIP's use of config headers to govern behavior, as there are alternatives that are less fragile (if possibly fatter).

**Comment by mspang** [comment]
> lgtm.
> 
> This would have saved me a lot of work in my recent "gn in chip_xcode_connector.sh" work, but I'm a little concerned longer term CHIP's use of config headers to govern behavior, as there are alternatives that are less fragile (if possibly fatter).

I share the concern. I like openthread's approach - C API that doesn't change based on config. Is that what you're thinking or something else?

File: examples/lighting-app/nrf5/BUILD.gn (hunk 1 above)
Comment (mspang) [reply examples/lighting-app/nrf5/BUILD.gn:17]: This is unrelated I just noticed it was missing from lighting app (it's in lock-app).

**Comment by rwalker-apple** [comment]
> > lgtm.
> > This would have saved me a lot of work in my recent "gn in chip_xcode_connector.sh" work, but I'm a little concerned longer term CHIP's use of config headers to govern behavior, as there are alternatives that are less fragile (if possibly fatter).
> 
> I share the concern. I like openthread's approach - C API that doesn't change based on config. Is that what you're thinking or something else?

Yes: an API that doesn't change at a minimum

**Review by e-rk — APPROVED** [review]
Works great. Thanks for this. :+1:

**Review by woody-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

**Pull request merged**